pom.xml

```diff
@@ -203,5 +203,11 @@
             <version>1.10.7</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>envinject</artifactId>
```

Comment:
This plugin is useless for Pipeline. No idea what you are trying to do with https://github.com/jenkinsci/pipeline-groovy-lib-plugin/blob/4ce1164e6ed9c9e0bf6689e07f73110789f1c099/src/test/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetrieverTest.java#L1058 but this is not the way to do it.

Reply:
(Maybe you wanted a `@TestExtention` on an `EnvironmentContributor` or something?)

Reply:
Thanks, feared so. 

As commented in the test, the idea was to see if we can have a Job or individual Run with some (externally) pre-set environment variables. I mean, any Run already does have a personal copy of EnvVars inherited from the global config or particular agent. Makes sense for there to be a way that somehow else we can intrude into what happens between the acts of Job definition and a particular build's start. 

I hoped to test that this works, e.g. for `@Library('libname@${env.VARNAME}')` the job-customized variable takes precedence - so I hoped `envinject` or some other plugin can be it. If envinject is not the way (or if Pipeline generally definitely would not make it possible), that's also good to know and I can then excise this test attempt. So far I left it there for someone smarter to complete - scaffolding is ready, just some magic spell missing.

Reply:
Thinking of it, the test code schedules a build and waits for outcomes (several ways to do so). Is there a programmatic way to do what `scheduleBuild` does but in more detailed explicit commands - e.g. to create that WorkflowRun object from a WorkflowJob, then somehow inject envvars (is the return value of `getEnvironment()` a copy of its `EnvVars`, or a reference to the real list - to which we can `add()`?) and just then schedule it to be built?

Reply:
Not sure we are discussing same thing in this part, or I misunderstood something? I did not find any tests to copy-paste from (here or in other repos) that would deal with setting envvars into pipeline Job or Run objects :\

The closest I came to test the added `@Library('libname@${env.VARNAME}')` feature here is setting an envvar Jenkins Global configuration and so the built-in agent. I assume defining a build agent and setting *its* custom envvars could be doable, but the scaffolding needed to create and connect a separate agent seemed prohibitively complex.

I did not find how to inject "custom" external envvars (at higher "priority" than those inherited from a build agent) into a pipeline, or certain info on whether it is possible or not.

Reply:
> inject "custom" external envvars

That would be `EnvironmentContributor`. My comment about existing tests was about testing actual multibranch—IOW rather than mocking up environment variables, actually have them be injected naturally by production code. From a quick search, https://github.com/jenkinsci/pipeline-groovy-lib-plugin/blob/8fb8c55868cc91791f7e5bc39c40594597329128/src/test/java/org/jenkinsci/plugins/workflow/libs/SCMRetrieverTest.java#L60-L74

Reply:
Finally got what you meant with `EnvironmentContributor` looking at https://github.com/jenkinsci/subversion-plugin/blob/master/src/test/java/hudson/scm/SubversionSCMTest.java#L1383-L1390

So far experiments are mixed: I can inject the envvar in Job and Run, but apparently something in https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/Run.java#L2441-L2444 overwrites back it from global server setting (after Job setting alone the value is as expected - it does not tap into `EnvironmentContributingAction`s).

Reply:
As for multibranch - my intention in the dozen added tests was to define and run various project and job types, effectively a matrix of "True MBP", "MBP with SingleSCM", a "plain pipeline from single SCM" and an "inline" pipeline with no SCM, multiplexed by use of static "version" names, expanded `libname@${BRANCH_NAME}` as suitable for the job definition, and a bit with `libname@{env.VARNAME}`.

So if that comment regarded defining `sampleRepo` for lib and/or the job via git - yes, I have a lot of that there. Introduced `sampleRepo2` to juggle two repos, where used.

Or did I miss the question again? :D

Reply:
As for `EnvironmentContributor` - I've tried to `add()` with different index locations, until I looked at https://github.com/jenkinsci/jenkins/blob/031f40c50899ec4e5fa4d886a1c006a5330f2627/core/src/main/java/hudson/ExtensionList.java#L296 and realized the index is ignored. "Hello table, meet the face..."

Reply:
What's the point of adding to the list, reversing the order for priorities... and then in `addSync()` doing a `sort()` of it every time? How do I ensure that my job/run-level contributor envvars overrides that of global config?

Reply:
I've tried even stashing the list contents to a temporary list, removing everything from original, adding my test entry and then adding back the original for a real-life feel - and no banana. Suppose `sort()` botches it (and then reverse or not in the queries at Job/Run `getEnvironment()` implems does not fully matter *if* the sorting order is not manageable).

Reply:
FWIW this adventure got me to realize that description of the feature for envvar-based choice of library is misleading (was to me too) - that something set on a build agent would impact it. Pipeline itself is only ever parsed on a "built-in" node, right?

The idea still does apply to running same pipelines on different Jenkins controllers though...

And maybe to EnvVar injection by plugins or other logic, if I get it to work reliably.

Reply:
Stepping through the call stack with IDEA I found (coulda deduced) that in this test case the https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L504 chimes in with an unique piece of logic different from that used by other plugins, to query `instance.getGlobalNodeProperties()` (and IMHO this one would belong higher in the stack to be consistent across the board). 

In my test cases, this overlays previously set global-config envvar over anything set by current build node (agent), various `EnvironmentContributor` implems from plugins or elsewhere, and `EnvironmentContributingAction`s -- all consulted by `super` and `getParent` calls to `Run` and `Job`, just because this build type is a `WorkflowRun` and looks at global configuration after all these preference setters.

Reply:
For kicks, also traced my implementation with envInject (the currently committed PR branch state) - it also seems to be consulted as an `Action`,  but fails to latch in this `EnvironmentContributingAction::buildEnvironment()` line (from IDEA decompiler):
````
if (run instanceof AbstractBuild && Util.isOverridden(EnvironmentContributingAction.class, this.getClass(), "buildEnvVars", new Class[]{AbstractBuild.class, EnvVars.class})) { ... }
````

From this I quickly guess envInject did not override "buildEnvVars" method and so is ignored. But otherwise `this` does contain my `envMap` contents and would have applied here (and probably lost to `WorkflowRun` overwriting it from global config, per above.

Also earlier in `getEnvironment` the plugin was consulted for `EnvInjectEnvVarsContributor::buildEnvironmentFor()` but skipped because it got `null` from the likes of:
````
EnvInjectJobProperty jobProperty = (EnvInjectJobProperty)job.getProperty(EnvInjectJobProperty.class);
````

Reply:
Pushed the recent changes to the test code (which so far fails, but seems the preferences of `WorkflowRun` detailed above explain why). I suppose I'd break the test into pieces to try such injection to prove that a plugin might technically influence the choice of `@Library` via envvars (although just now it would not be envinject, it seems). Might keep the current test code in place to confirm that global-config envvars always land on top, now that it is sort of expected.

Reply:
Added the test update regarding "injected" envvars vs. global config - PR updated.
Hope to find how to manage the "built-in" node config and add its envvars to the realistic use-cases modeled by the test case.

Reply:
> As for `EnvironmentContributor` - I've tried to `add()`…

No, do not call `add`, use the `@TestExtension` annotation as previously mentioned.

Reply:
> > As for `EnvironmentContributor` - I've tried to `add()`…
> No, do not call `add`, use the `@TestExtension` annotation as previously mentioned.

Sorry, did not wrap my head around those (too advanced Java I guess), so posted what works for me :)

Updated the PR now with tests manipulating the "built-in" node's envvar configuration, with the variable NOT set in global config, although this did not work the way I expected it to:
* per (fix to) https://issues.jenkins.io/browse/JENKINS-27739 envvars are cached when first queried from a `Computer.getEnvironment()` and the only way to clear the cache is to reconnect the agent (after the fix), delete and redefine it (before the fix), or restart Jenkins. Neither of these apply to "built-in" node (can't disconnect or delete) and the JenkinsRule instance respectively.
* I did manage to use Java reflection to set that `cachedEnvironment` field to `null` and checked it with the debugger - it was repopulated from the `getChannel()` which for `MasterComputer` ended up reading global config.
* I also verified with a separate Jenkins instance - envvars set in "built-in" node configuration are ignored even if I restart the instance. Arguably a bug (either don't let them be configured, or do honour them) but one for another day.

So effectively the updated end of test currently verifies just that - the expected behavior is what it is (and not so much what I naively wanted it to be); but now we would be alarmed if the core changes (fixes) it and this test can get updated accordingly.